documentation/src/main/asciidoc/topics/search.adoc

```diff
@@ -0,0 +1,1096 @@
+
+[source,xml,options="nowrap",subs=attributes+]
+----
+include::config_examples/indexing_local_heap.xml[]
+----
+
+[[query_index_manager]]
+=== Index Manager
+
+{brandname} uses internally a component called "Index Manager" to control how new data is applied to the index and
+when the data is visible to searches.
+
+The default Index Manager `directory-based` writes to the index as soon as the data is written to the cache. The downside is it can slow down considerably cache writes specially under heavy writing scenarios, since it needs to do constant costly operations called "flushes" on the index.
+
+The `near-real-time` index manager is similar to the default index manager but takes advantage of the Near-Real-Time features of Lucene. It has better write performance because it flushes the index to the underlying store less often.
+The drawback is that unflushed index changes can be lost in case of a non-clean shutdown. Can be used in conjunction with `local-heap` or `filesystem`.
+
+Example with `local-heap`:
+
+[source,xml,options="nowrap",subs=attributes+]
+----
+include::config_examples/indexing_near_real_time.xml[]
+----
+
+Example with `filesystem`:
+
+[source,xml,options="nowrap",subs=attributes+]
+----
+include::config_examples/indexing_near_real_time_fs.xml[]
+----
+
+[[query_massindexer]]
+=== Rebuilding Indexes
+
+Rebuilding an index reconstructs it from data stored in the cache. You need to rebuild indexes if you change things like
+definitions of indexed types or Analyzers. Likewise you might need to rebuild indexes if they are deleted for some reason.
+Beware it might take some time as it needs to reprocess all data in the grid!
+
+[source,java]
+----
+Indexer indexer = Search.getIndexer(cache);
+CompletionStage<Void> future = index.run();
+----
+
+//TODO
+//[[query_indexless]]
+//==== Indexless
+
+//TODO
+//[[query_hybrid]]
+//==== Hybrid
+
+== Searching
+
+Create relational and full-text queries in both Library and Remote Client-Server mode with the Ickle query language.
+
+To use the API, first obtain a QueryFactory to the cache and then call the `.create()` method, passing in the string
+to use in the query. Each _QueryFactory_ instance is bound to the same _Cache_ instance as the _Search_, but it is
+otherwise a stateless and thread-safe object that can be used for creating multiple queries in parallel.
+
+For instance:
+
+[source,java,tile="Using Ickle"]
+----
+// Remote Query, using protobuf
+QueryFactory qf = org.infinispan.client.hotrod.Search.getQueryFactory(remoteCache);
+Query q = qf.create("from sample_bank_account.Transaction where amount > 20");
+
+// Embedded Query using Java Objects
+QueryFactory qf = org.infinispan.query.Search.getQueryFactory(cache);
+Query q = qf.create("from com.acme.Book where price > 20");
+
+// Execute the query
+QueryResult<Book> queryResult = q.execute();
+----
+
+
+NOTE: A query will always target a single entity type and is evaluated over the contents of a single cache. Running a
```

Comment:
Notes, tips, and warnings should follow the format defined here https://infinispan.org/docs/11.0.x/titles/contributing/contributing.html#documentation_guidelines_format_reference

```
[NOTE]
====
This is a note that you should read.
====
```